<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [29]:
import pandas as pd
import psycopg2

In [30]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port=5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [3]:
# текст запроса
query_3_1 = f'''select
    distinct
    count(id) result
from
    public.vacancies
'''

In [35]:
# результат запроса
result_query_3_1 = pd.read_sql_query(query_3_1, connection)
result_query_3_1

C:\Users\grish\AppData\Local\Temp\ipykernel_2452\2526852126.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_query_3_1 = pd.read_sql_query(query_3_1, connection)


result
0   49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [32]:
# текст запроса
query_3_2 = f'''select
    distinct
    count(id) result
from
    public.employers
'''

In [33]:
# результат запроса
result_query_3_2 = pd.read_sql_query(query_3_2, connection)
result_query_3_2

C:\Users\grish\AppData\Local\Temp\ipykernel_2452\7577267.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_query_3_2 = pd.read_sql_query(query_3_2, connection)


result
0   23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [30]:
# текст запроса
query_3_3 = f'''select
    distinct
    count(id) result
from
    public.areas
'''

In [31]:
# результат запроса
result_query_3_3 = pd.read_sql_query(query_3_3, connection)
result_query_3_3

C:\Users\grish\AppData\Local\Temp\ipykernel_2452\2033642199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_query_3_3 = pd.read_sql_query(query_3_3, connection)


result
0    1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [36]:
# текст запроса
query_3_4 = f'''select
    distinct
    count(id) result
from
    public.industries'''

In [37]:
# результат запроса
result_query_3_4 = pd.read_sql_query(query_3_4, connection)
result_query_3_4

C:\Users\grish\AppData\Local\Temp\ipykernel_2452\1962420110.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_query_3_4 = pd.read_sql_query(query_3_4, connection)


result
0     294

***

Всего в базе находится 49197 уникальных вакансий в 294 сферах деятельности. Вакансии предлагают больше 23,5 тысяч работодателей из 1362 регионов.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [57]:
# текст запроса
vac_cnt = f'''
select
    a.name "Регион",
    count(v.id) "Количество вакансий"
from public.vacancies v
join public.areas a on v.area_id = a.id
group by 1
order by 2 desc
limit 5'''

In [58]:
# результат запроса
result_vac_cnt = pd.read_sql_query(vac_cnt, connection)
result_vac_cnt

C:\Users\grish\AppData\Local\Temp\ipykernel_2452\1054740229.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_vac_cnt = pd.read_sql_query(vac_cnt, connection)


Регион  Количество вакансий
0           Москва                 5333
1  Санкт-Петербург                 2851
2            Минск                 2112
3      Новосибирск                 2006
4           Алматы                 1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [38]:
# текст запроса
salary = f'''select
    salary_from,
    salary_to
from public.vacancies
where (salary_from is not null) or (salary_to is not null)
'''

In [39]:
# результат запроса
salary_result = pd.read_sql_query(salary, connection)
salary_result

C:\Users\grish\AppData\Local\Temp\ipykernel_11476\3240231440.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_result = pd.read_sql_query(salary, connection)


salary_from  salary_to
0          64000.0        NaN
1          15000.0        NaN
2          40000.0    60000.0
3          50000.0        NaN
4         150000.0        NaN
...            ...        ...
24068      50000.0        NaN
24069     102500.0        NaN
24070      40000.0    78000.0
24071      60000.0        NaN
24072     146456.0   244094.0

[24073 rows x 2 columns]

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [63]:
# текст запроса
salary_avg = f'''select
    round(avg(salary_from)) lower_salary_average,
    round(avg(salary_to)) upper_salary_average
from
    public.vacancies'''

In [64]:
# результат запроса
result_salary_avg = pd.read_sql_query(salary_avg, connection)
result_salary_avg

C:\Users\grish\AppData\Local\Temp\ipykernel_2452\1322697596.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_salary_avg = pd.read_sql_query(salary_avg, connection)


lower_salary_average  upper_salary_average
0               71065.0              110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
employment_schedule = f'''
select
    distinct
    schedule,
    employment,
    count(*)
from
    public.vacancies
group by 1,2
order by 3 desc
'''

In [19]:
# результат запроса
result_empl_sch = pd.read_sql_query(employment_schedule, connection)
result_empl_sch

C:\Users\grish\AppData\Local\Temp\ipykernel_2064\1663252333.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_empl_sch = pd.read_sql_query(employment_schedule, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [24]:
# текст запроса
experience = f'''
select
    distinct
    experience,
    count(*)
from
    public.vacancies
group by 1
order by 2'''

In [25]:
# результат запроса
result_experience = pd.read_sql_query(experience, connection)
result_experience

C:\Users\grish\AppData\Local\Temp\ipykernel_2064\3159975750.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_experience = pd.read_sql_query(experience, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Наибольшее количество вакансий предлагают в Москве (5333 вакансии). В пятёрку лидеров по этому показателю также входят Санкт-Петербург, Минск, Новосибирск и Алматы.

В 24072 из 49197 вакансий заполнено хотя бы одно из двух полей с зарплатой. Средние значения для нижней и верхней зарплатных вилок составляют 71065 руб. и 110537 руб.

Наиболее часто работодатели предлагают полный рабочий день и полную занятость — таких вакансий 35367 из 49197. Также в тройку наиболее популярного рабочего графика входят удалённая работа и гибкий график. 

Наиболее часто встречаются вакансии с требуемым опытом работы от 1 года до 3 лет (почти половина вакансий). Наименее часто — с опытом работы более 6 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [30]:
# текст запроса
employers = f'''
select
    e.name,
    count(v.id)
from
    public.employers e
join
    public.vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
limit 5
'''

In [31]:
# результат запроса
result_employers = pd.read_sql_query(employers, connection)
result_employers

C:\Users\grish\AppData\Local\Temp\ipykernel_2064\3226855445.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_employers = pd.read_sql_query(employers, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [15]:
# текст запроса
area_employers = '''
select
    a.name,
    count(v.id) vac_cnt,
    count(e.id) empl_cnt
from
    public.areas a
left join
    public.employers e on a.id = e.area
left join
    public.vacancies v on a.id = v.area_id
where v.id is null
group by 1
order by 3 desc
limit 5
'''

In [16]:
# результат запроса
result_area_employers = pd.read_sql_query(area_employers, connection)
result_area_employers

C:\Users\grish\AppData\Local\Temp\ipykernel_11500\508582234.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_area_employers = pd.read_sql_query(area_employers, connection)


name  vac_cnt  empl_cnt
0              Россия        0       410
1           Казахстан        0       207
2  Московская область        0        75
3  Краснодарский край        0        19
4            Беларусь        0        18

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [31]:
# текст запроса
employers_area_count = f'''
select
    e.name,
    count(distinct v.area_id) "количество регионов"
from
    public.employers e
left join
    public.vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
'''

In [32]:
# результат запроса
result_empl_area = pd.read_sql_query(employers_area_count, connection)
result_empl_area

C:\Users\grish\AppData\Local\Temp\ipykernel_11500\538426741.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_empl_area = pd.read_sql_query(employers_area_count, connection)


name  количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
23170                  СДЕЛКА                    0
23171            Alandr Group                    0
23172                     СДК                    0
23173      Lemon Land Lombard                    0
23174                  Махина                    0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [8]:
# текст запроса
employers_industries = f'''
select
    count(e.id) result
from
    public.employers e
left join
    public.employers_industries ei on ei.employer_id = e.id
where ei.industry_id is null
'''

In [9]:
# результат запроса
result_employers_industries = pd.read_sql_query(employers_industries, connection)
result_employers_industries

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\3395158327.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_employers_industries = pd.read_sql_query(employers_industries, connection)


result
0    8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [10]:
# текст запроса
employers_industries_3_4 = f'''
select
    e.name "Название компании",
    count(ei.industry_id) "Количество сфер деятельности"
from 
    public.employers e
join
    public.employers_industries ei on e.id = ei.employer_id
group by 1
having count(ei.industry_id) = 4
order by 1
offset 2
limit 1
'''

In [11]:
# результат запроса
result_employers_industries_3_4 = pd.read_sql_query(employers_industries_3_4, connection)
result_employers_industries_3_4

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\260977877.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_employers_industries_3_4 = pd.read_sql_query(employers_industries_3_4, connection)


Название компании  Количество сфер деятельности
0              2ГИС                             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [12]:
# текст запроса
software_development = f'''
select
    i.name "Cфера деятельности",
    count(e.employer_id) "Количество работодателей"
from
    employers_industries e
left join
    industries i on e.industry_id = i.id
where i.name = 'Разработка программного обеспечения'
group by 1
'''

In [13]:
# результат запроса
result_software_development = pd.read_sql_query(software_development, connection)
result_software_development

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\3529071323.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_software_development = pd.read_sql_query(software_development, connection)


Cфера деятельности  Количество работодателей
0  Разработка программного обеспечения                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [15]:
# код для получения списка городов-милионников
import urllib.parse as req
import pandas as pd

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
df = pd.read_html(url)[1]
cities = tuple(df['Город'].tolist())
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [16]:
# текст запроса
area_vacancies = f'''
(select
    a.name "Регион",
    count(v.id) "Количество вакансий"
from
    public.areas a
left join
    public.vacancies v on a.id = v.area_id
join
    public.employers e on v.employer_id = e.id
where e.name = 'Яндекс' and a.name in {cities}
group by 1
order by 2)

union all

select
    'Total',
    sum(x.count)
from
    (select
    count(v.id) as count
from
    public.areas a
left join
    public.vacancies v on a.id = v.area_id
join
    public.employers e on v.employer_id = e.id
where e.name = 'Яндекс' and a.name in {cities}) as x

order by 2 desc
'''

In [17]:
# результат запроса
result_area_vacancies = pd.read_sql_query(area_vacancies, connection)
result_area_vacancies

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\1390180641.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_area_vacancies = pd.read_sql_query(area_vacancies, connection)


Регион  Количество вакансий
0             Total                485.0
1            Москва                 54.0
2   Санкт-Петербург                 42.0
3      Екатеринбург                 39.0
4   Нижний Новгород                 36.0
5       Новосибирск                 35.0
6           Воронеж                 32.0
7         Краснодар                 30.0
8            Самара                 26.0
9               Уфа                 26.0
10            Пермь                 25.0
11   Ростов-на-Дону                 25.0
12           Казань                 25.0
13        Волгоград                 24.0
14       Красноярск                 23.0
15        Челябинск                 22.0
16             Омск                 21.0

***

Больше всего вакансий (1933) предлагает компания Яндекс. В пятёрку лидеров по этому признаку также входят Ростелеком, Тинькофф, Сбер и Газпром Нефть.

Россия является регионом, где наблюдается наибольшее количество работодателей при отсутствии вакансий.

Компания, вакансии которой присутствуют в максимальном количестве регионов — это Яндекс (предложения в 181 регионах). На втором месте — Ростелеком (152 региона)

В вакансиях от 8419 работодателей (~1/5 от общего количества) не указана сфера деятельности

3553 работодателя предлагают вакансии в сфере деятельности, связанной с программным обеспечением

Компания Яндекс предлагает кандидатам 485 вакансий в городах миллионниках. Наибольшее количество вакансий (54) в Москве.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [22]:
# текст запроса
data_vacancies = f'''
select 
    count(*) result
from public.vacancies
where 
    name like '%data%' 
    or name like '%данн%'
    or name like '%Data%'
    or name like '%DATA%'
    or name like '%Данн%'
    or name like '%ДАНН%'
'''

In [23]:
# результат запроса
result_data_vacancies = pd.read_sql_query(data_vacancies, connection)
result_data_vacancies

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\1482277660.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_data_vacancies = pd.read_sql_query(data_vacancies, connection)


result
0    1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [24]:
# текст запроса
junior_vacancies = '''
select
    count(*) result
from
    public.vacancies
where
    (name ilike any ('
	{%data scientist%, %data science%, %исследователь данных%, 
	%machine learning%, %machine learning%, %машинн%обучен%}
	') or 
	(name not ilike '%HTML%' and name like '%ML%'))
and
    (lower(name) like '%junior%'
    or
    experience = 'Нет опыта'
    or
    lower(employment) like '%Стажировка%')
'''

In [25]:
# результат запроса
result_junior_vacancies = pd.read_sql_query(junior_vacancies, connection)
result_junior_vacancies

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\1768022047.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_junior_vacancies = pd.read_sql_query(junior_vacancies, connection)


result
0      51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [26]:
# текст запроса
sql_postgres_vacancies = '''
select
    count(*) result
from
    public.vacancies
where
    (name ilike any ('
	{%data scientist%, %data science%, %исследователь данных%, 
	%machine learning%, %machine learning%, %машинн%обучен%}
	') or 
	(name not ilike '%HTML%' and name like '%ML%'))
and
    (key_skills ilike '%SQL%' 
    or key_skills like '%postgres%')
 '''

In [27]:
# результат запроса
result_sql_postgres_vacancies = pd.read_sql_query(sql_postgres_vacancies, connection)
result_sql_postgres_vacancies

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\3715360399.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_sql_postgres_vacancies = pd.read_sql_query(sql_postgres_vacancies, connection)


result
0     201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [28]:
# текст запроса
python_vacancies = '''
select
    count(*) result
from
    public.vacancies
where
    (name ilike any ('
	{%data scientist%, %data science%, %исследователь данных%, 
	%machine learning%, %machine learning%, %машинн%обучен%}
	') or 
	(name not ilike '%HTML%' and name like '%ML%'))
 and
	(key_skills ilike '%python%')
 '''

In [29]:
# результат запроса
result_python_vacancies = pd.read_sql_query(python_vacancies, connection)
result_python_vacancies

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\3235981748.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_python_vacancies = pd.read_sql_query(python_vacancies, connection)


result
0     351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [32]:
# текст запроса
key_skills_DS = '''
select
    round(AVG((LENGTH(v.key_skills) - LENGTH(replace(v.key_skills,CHR(9),'')))::integer+1),2) result
from
    public.vacancies v
where
    (name ilike any ('
	{%data scientist%, %data science%, %исследователь данных%, 
	%machine learning%, %machine learning%, %машинн%обучен%}
	') or 
	(name not ilike '%HTML%' and name like '%ML%'))
	and v.key_skills is not null
 '''

In [33]:
# результат запроса
result_key_skills_DS = pd.read_sql_query(key_skills_DS, connection)
result_key_skills_DS

C:\Users\grish\AppData\Local\Temp\ipykernel_9624\2459272466.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_key_skills_DS = pd.read_sql_query(key_skills_DS, connection)


result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [38]:
# текст запроса
avg_salary_data = '''
select
    v.experience "Опыт работы",
    round(avg(coalesce((v.salary_to+v.salary_from)/2, v.salary_from, v.salary_to)),0)"Средний уровень ЗП (руб.)"
from
    public.vacancies v
where
    (name ilike any ('
	{%data scientist%, %data science%, %исследователь данных%, 
	%machine learning%, %machine learning%, %машинн%обучен%}
	') or 
	(name not ilike '%HTML%' and name like '%ML%'))
	and v.key_skills is not null
group by 1
order by 2 desc
    '''

In [39]:
# результат запроса
result_avg_salary_data = pd.read_sql_query(avg_salary_data, connection)
result_avg_salary_data

C:\Users\grish\AppData\Local\Temp\ipykernel_2212\2184833925.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_avg_salary_data = pd.read_sql_query(avg_salary_data, connection)


Опыт работы  Средний уровень ЗП (руб.)
0         Более 6 лет                        NaN
1       От 3 до 6 лет                   243115.0
2  От 1 года до 3 лет                   146704.0
3           Нет опыта                    74643.0

***

1771 вакансия из датасета имеет отношение к данным. Из них 51 подойдёт для начинающего специалиста

В датасете встречается 201 вакансия для DS, где в качестве ключевого навыка указан SQL или postgres и 351 вакансия, где в качестве ключевого навыка указан python

В среднем в вакансиях для DS работодатели указывают 6 ключевых навыков

Data Scientist без опыта работы может рассчитывать на среднюю зарплату в 74,6 тыс. руб., специалисты с опытом работы от 1 года до 3 лет получают ~146,7 тыс. руб., а со стажем выше 3-х лет — в среднем 243,1 тыс. руб.

# Общий вывод по проекту

In [18]:
general_query = '''
select
    i.name "Название сферы деятельности",
    count(v.id) "Количество вакансий"
from
    public.vacancies v
join
    public.employers_industries ei on v.employer_id = ei.employer_id
join    
    public.industries i on ei.industry_id = i.id
group by 1
order by 2 desc
limit 10
    '''

In [19]:
result_general_query = pd.read_sql_query(general_query, connection)
result_general_query

C:\Users\grish\AppData\Local\Temp\ipykernel_2212\2552191818.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_general_query = pd.read_sql_query(general_query, connection)


Название сферы деятельности  Количество вакансий
0                Разработка программного обеспечения                12499
1  Системная интеграция,  автоматизации технологи...                11034
2  Интернет-компания (поисковики, платежные систе...                 6413
3                                               Банк                 2742
4                                    Мобильная связь                 1585
5                                 Кадровые агентства                 1143
6                              Консалтинговые услуги                 1137
7                                 Интернет-провайдер                 1095
8                       Розничная сеть (продуктовая)                 1034
9                                Фиксированная связь                 1028

In [31]:
general_query_1 = '''
select
    i.name "Название сферы деятельности",
    count(v.id) "Количество вакансий"
from
    public.vacancies v
join
    public.employers_industries ei on v.employer_id = ei.employer_id
join    
    public.industries i on ei.industry_id = i.id
where
    (v.name ilike any ('
	{%data scientist%, %data science%, %исследователь данных%, 
	%machine learning%, %machine learning%, %машинн%обучен%}
	') or 
	(v.name not ilike '%HTML%' and v.name like '%ML%'))
group by 1
order by 2 desc
limit 10
'''

In [32]:
result_general_query_1 = pd.read_sql_query(general_query_1, connection)
result_general_query_1

C:\Users\grish\AppData\Local\Temp\ipykernel_2212\1602942.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_general_query_1 = pd.read_sql_query(general_query_1, connection)


Название сферы деятельности  Количество вакансий
0                Разработка программного обеспечения                  148
1  Системная интеграция,  автоматизации технологи...                  139
2                                               Банк                   92
3  Интернет-компания (поисковики, платежные систе...                   79
4                                    Мобильная связь                   23
5  Маркетинговые, рекламные, BTL, дизайнерские, E...                   22
6  Управляющая, инвестиционная компания (управлен...                   21
7                              Консалтинговые услуги                   16
8                                Фиксированная связь                   13
9  Научно-исследовательская, научная, академическ...                    9

In [44]:
general_query_2 = '''
select
    i.name "Сфера деятельности",
    round(avg(coalesce((v.salary_from+v.salary_to)/2, v.salary_from, v.salary_to)),0) "Средний уровень ЗП"
from
    public.vacancies v
join
    public.employers_industries ei on v.employer_id = ei.employer_id
join    
    public.industries i on ei.industry_id = i.id
group by 1
order by 2 desc
limit 20
'''

In [45]:
result_general_query_2 = pd.read_sql_query(general_query_2, connection)
result_general_query_2

C:\Users\grish\AppData\Local\Temp\ipykernel_2212\2935411121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_general_query_2 = pd.read_sql_query(general_query_2, connection)


Сфера деятельности  Средний уровень ЗП
0                                           Автошкола                 NaN
1              Ассоциация в сфере культуры, искусства                 NaN
2           Агрохимия (продвижение, оптовая торговля)            381429.0
3                            Агрохимия (производство)            339875.0
4                                      Игорный бизнес            191905.0
5                   Корма для животных (производство)            168300.0
6                                          Зоомагазин            167613.0
7                                  Кадровые агентства            144530.0
8       Лакокрасочная продукция, сырье (производство)            142803.0
9   Бытовая химия, парфюмерия, косметика  (произво...            130120.0
10  Лакокрасочная продукция, сырье (продвижение, о...            128445.0
11  Дорожно-строительная техника, сельскохозяйстве...            122878.0
12  Дорожно-строительная техника, сельскохозяйстве...            119919.0
13  Оборудование для пищевой промышленности, упако...            119656.0
14  Финансово-кредитное посредничество (биржа, бро...            117714.0
15                Разработка программного обеспечения            117659.0
16  Управляющая, инвестиционная компания (управлен...            114297.0
17  Корма для животных (продвижение, оптовая торго...            113667.0
18                          Ветеринарная деятельность            113364.0
19                     Учреждение соц.помощи и защиты            113125.0

По результатам проведённого исследования я вывяснила, что на данный момент наибольшее количество вакансий актуально в сфере разработки программного обеспечения - это 12499 вакансий из 49197 ~ 1/4 от общего количества. В пятёрку сфер с наибольшим количеством рабочих мест также входят системная интеграция, интернет-компании, банки и мобильная связь

Эти же сферы деятельности являются наиболее популярными для специалистов по DS

Наиболее популярная сфера деятельности для специалистов по DS - разработка программного обеспечения. Эта сфера находится на 15-м месте в рейтинге сфер деятельности с наибольшей ЗП. Наиболее оплачиваемыми сферами деятельности являются агрохимия, игорный бизнес и бизнес, связанный с животными (производство кормов для животных, зоомагазины)

Обобщая выводы, сделанные по блокам, можно сделать утверждение, что специалист в области машинного обучения - это востребованная и высокооплачиваемая профессия. А знания, которыми обладают DS, могут пригодиться в любой области